controllers/vspherecluster_controller.go

```diff
@@ -321,6 +333,49 @@ func (r *VSphereClusterReconciler) reconcileCloudProvider(ctx *context.ClusterCo
 }
 
 func (r *VSphereClusterReconciler) reconcileStorageProvider(ctx *context.ClusterContext) error {
+	// if storage config is not defined, assume we don't want CSI installed
+	storageConfig := ctx.VSphereCluster.Spec.CloudProviderConfiguration.ProviderConfig.Storage
+	if storageConfig == nil {
+		ctx.Logger.V(2).Info(
+			"storage config was not specified in VSphereCluster, skipping reconciliation of the CSI driver",
+			"cluster", ctx.ClusterNamespacedName(),
+		)
+
+		return nil
+	}
+
+	// if at least 1 field in the storage config is defined, assume CNS should be installed
+	// and use default images when not defined
+	if storageConfig.ControllerImage == "" {
+		storageConfig.ControllerImage = cloudprovider.DefaultCSIControllerImage
+	}
+
+	if storageConfig.NodeDriverImage == "" {
```

Comment:
@andrewsykim  - maybe at some point we should consider shipping webhooks for both high-level validation and defaulting

Reply:
HA! I was literally just mentioning this to Andrew yesterday :)

Reply:
I’ll open an issue for this one.